### DISCUSSION TITLE
How to implement two-way communication between virtual machines?

### DISCUSSION THREAD

**Discussion by Sunt-git** [bao-project/bao-hypervisor#31] [General]
I ran it according to the demo, linux can actively communicate with freertos, but how does freertos actively communicate with linux?

**Comment by josecm** [comment]
The communication channel works the same in both directions. 

In the provided demos,  Linux can read the /dev/baoipc0 device to get FreeRTOS messages from the shared memory. However, we haven't written the Linux driver to enable and receive the interrupts, but this is essentially what's missing for Linux to receive asynchronous notifications from the FreeRTOS side.
